### ISSUE TITLE
Why Batch inference is not support currently？

### ISSUE THREAD

**Issue by LovingThresh** [open-mmlab/mmsegmentation#2644]
AssertionError: Batch inference is not support currently, as the image size might be different in a batch.

It is Confusing.

**Comment by xiexinch** [comment]
Hi @LovingThresh,
Since the validation or test set images have different shapes, support batch inference at evaluation might affect the result.

**Comment by louan1998** [comment]
I also encountered this problem. I ran this module "unet_s5-d16_deeplabv3_4xb4-40k_chase-db1-128x128.py" following the official documentation of MMSegmentation to train the retina dataset, and reported this error when it reached 4000 iterations. This officially supported retina The image size of the data set is the same, why?

**Comment by louan1998** [comment]
> I also encountered this problem. I ran this module "unet_s5-d16_deeplabv3_4xb4-40k_chase-db1-128x128.py" following the official documentation of MMSegmentation to train the retina dataset, and reported this error when it reached 4000 iterations. This officially supported retina The image size of the data set is the same, why?

Then I continued to use the "--resume" command to continue training, but after 8000 iterations, the error "**AssertionError: Batch inference is not supported currently, as the image size might be different in a batch**" was prompted again. It seems that this error will occur as long as it iterates 4000 times.

**Comment by louan1998** [comment]
I followed the official documentation to run the "unet_s5-d16_deeplabv3_4xb4-40k_chase-db1-128x128.py" module, and reported an error "AssertionError: Batch inference is not support currently, as the image size might be different in a batch" when iterating to 4000 times, and then I use the --resume command to continue training on the weight of 4000 iterations, and report this error again when it reaches 8000 times. This data set is also supported by mmseg. I saw that the size of the pictures inside is the same. why is that?

**Comment by louan1998** [comment]
@edsml-hmc122 That's right! I noticed that too! If batch_size = 1 in **val_dataloader** ，there is no problem!

**Comment by chenhuagg** [comment]
@louan1998 How did you solve it in the end?

**Comment by louan1998** [comment]
> @louan1998 How did you solve it in the end?

Set the batch_size in the val_dataloader to 1
![1690684900621](https://github.com/open-mmlab/mmsegmentation/assets/52094628/82c7dd20-68ce-4015-ae3f-95cea72b05f9)

**Comment by xiexinch** [comment]
Hi @edsml-hmc122,
I removed this limitation, you might have a try. If there are any problems, feel free to create an issue, and we'll fix it asap.
 https://github.com/open-mmlab/mmsegmentation/blob/main/mmseg/models/data_preprocessor.py#L135

**Comment by xiexinch** [comment]
> @xiexinch
> 
> > Hi @edsml-hmc122, I removed this limitation, you might have a try. If there are any problems, feel free to create an issue, and we'll fix it asap. https://github.com/open-mmlab/mmsegmentation/blob/main/mmseg/models/data_preprocessor.py#L135
> 
> Hi, I have done some testing. Seems that in the general sense, batched inference is working. More GPU VRAM is used and the logger shows fewer total iterations (since batches are larger).
> 
> However, the larger I make the batch size, the lower also the inference speed, so it doesn't lead to much acceleration from what I can tell. The GPU usage is also very strange, it is mostly 0 and spikes to 100% every 1-2 seconds. So I think this is a good start, but it doesn't seem to really accelerate the inference process yet.
> 
> Tested on an RTX 3080, using a batch size of 20 it used 5637MiB / 10240MiB VRAM. It took 1m22s for 1000 samples of size 512x512. Edit: With a batch size of 10, it took only ~20s.
> 
> Using a batch size of 1, the time was around 2m30s, I thought maybe the speedup should be larger.
> 
> Please re-open this issue if you think it's worth investigating.

Thanks for your feedback! We'll test this case and then find a better solution. Could you provide your config to us if it's available?
